// pigments/lib/color-expression.js
(function() {
  var ColorExpression, Q;

  Q = require('q');

  module.exports = ColorExpression = (function() {
    function ColorExpression(name, regexp, handle, priority) {
      this.name = name;
      this.regexp = regexp;
      this.handle = handle;
      this.priority = priority != null ? priority : 0;
      this.onigRegExp = new RegExp("^" + this.regexp + "$", 'i');
    }

    ColorExpression.prototype.canHandle = function(expression) {
      return this.onigRegExp.test(expression);
    };

    ColorExpression.prototype.searchSync = function(text, start) {
      var lastIndex, match, range, re, results, _ref;
      if (start == null) {
        start = 0;
      }
      results = void 0;
      re = new RegExp(this.regexp, 'gi');
      re.lastIndex = start;
      if (_ref = re.exec(text), match = _ref[0], _ref) {
        lastIndex = re.lastIndex;
        range = [lastIndex - match.length, lastIndex];
        results = {
          range: range,
          match: text.slice(range[0], range[1])
        };
      }
      return results;
    };

    ColorExpression.prototype.search = function(text, start, callback) {
      var defer, match, range, re, results, _ref;
      if (start == null) {
        start = 0;
      }
      if (callback == null) {
        callback = function() {};
      }
      defer = Q.defer();
      re = new RegExp(this.regexp, 'gi');
      if (!(_ref = re.exec(text), match = _ref[0], _ref)) {
        defer.resolve();
        callback();
        return;
      }
      range = [match.start, match.end];
      results = {
        range: range,
        match: text.slice(range[0], range[1])
      };
      defer.resolve(results);
      callback(results);
      return defer.promise;
    };

    return ColorExpression;

  })();

}).call(this);
